src/Http/Routing/src/Matching/HeaderMatcherPolicy.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal sealed class HeaderMatcherPolicy : MatcherPolicy, IEndpointComparerPolicy, IEndpointSelectorPolicy
+    {
+        /// <inheritdoc/>
+        // Run after HttpMethodMatcherPolicy (-1000) and HostMatcherPolicy (-100), but before default (0)
+        public override int Order => -50;
+
+        /// <inheritdoc/>
+        public IComparer<Endpoint> Comparer => new HeaderMetadataEndpointComparer();
+
+        /// <inheritdoc/>
+        bool IEndpointSelectorPolicy.AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            _ = endpoints ?? throw new ArgumentNullException(nameof(endpoints));
+
+            // When the node contains dynamic endpoints we can't make any assumptions.
+            if (ContainsDynamicEndpoints(endpoints))
+            {
+                return true;
+            }
+
+            return AppliesToEndpointsCore(endpoints);
+        }
+
+        /// <inheritdoc/>
+        public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
+        {
+            _ = httpContext ?? throw new ArgumentNullException(nameof(httpContext));
+            _ = candidates ?? throw new ArgumentNullException(nameof(candidates));
+
+            for (int i = 0; i < candidates.Count; i++)
+            {
+                if (!candidates.IsValidCandidate(i))
+                {
+                    continue;
+                }
+
+                var metadata = candidates[i].Endpoint.Metadata.GetMetadata<IHeaderMetadata>();
+                var metadataHeaderName = metadata?.HeaderName;
+                if (string.IsNullOrEmpty(metadataHeaderName))
```

Comment:
I would just key this based on metadata.Count == 0; Adding "" as the header name should be considered an error and we should reject the request, otherwise we run the risk of someone making a mistake and letting a request go through, Id rather be conservative here than risk an invalid match that accept a request that can have unwanted side-effects.